packages/core-sdk/resolvers/token-resolvers/resolvers/yearn/src/query/__tests__/e2e/integration.spec.ts

```diff
@@ -0,0 +1,147 @@
+      testEnvState.ipfs,
+      testEnvState.ensAddress,
+    );
+    client = new Web3ApiClient(clientConfig);
+  });
+
+  beforeAll(async () => {
+    // deploy api
+    const apiPath: string = path.join(path.resolve(__dirname), "../../../../");
+    const api = await buildAndDeployApi(apiPath, testEnvState.ipfs, testEnvState.ensAddress);
+    protocolEnsUri = `ens/testnet/${api.ensDomain}`;
+
+    // deploy token defiwrapper
+    const tokenApiPath: string = path.join(apiPath, "../../../../", "token");
+    const tokenApi = await buildAndDeployApi(
+      tokenApiPath,
+      testEnvState.ipfs,
+      testEnvState.ensAddress,
+    );
+    tokenEnsUri = `ens/testnet/${tokenApi.ensDomain}`;
+  });
+
+  afterAll(async () => {
+    await stopTestEnvironment();
+  });
+
+  describe("isValidProtocolToken", () => {
+    test("yearn_vault_v2 yvWBTC", async () => {
+      const result = await isValidProtocolToken(
+        v2_yvWBTC,
+        "yearn_vault_v2",
+        protocolEnsUri,
+        client,
+      );
+      expect(result.error).toBeFalsy();
+      expect(result.data).not.toBeUndefined();
+      expect(result.data).toBe(true);
+    });
+
+    test("yearn_vault_v1 yCrv3", async () => {
+      const result = await isValidProtocolToken(v1_y3Crv, "yearn_vault_v1", protocolEnsUri, client);
+      expect(result.error).toBeFalsy();
+      expect(result.data).not.toBeUndefined();
+      expect(result.data).toBe(true);
+    });
+
+    test("yearn_vault_v2 invalid protocol token", async () => {
+      const result = await isValidProtocolToken("0x1", "yearn_vault_v2", protocolEnsUri, client);
+      expect(result.error).toBeFalsy();
+      expect(result.data).not.toBeUndefined();
+      expect(result.data).toBe(false);
+    });
+
+    test("yearn_vault_v1 invalid protocol token", async () => {
+      const result = await isValidProtocolToken("0x1", "yearn_vault_v1", protocolEnsUri, client);
+      expect(result.error).toBeFalsy();
+      expect(result.data).not.toBeUndefined();
+      expect(result.data).toBe(false);
+    });
+  });
+
+  describe("getTokenComponents", () => {
+    test("yearn_vault_v2 yvWBTC", async () => {
+      const WBTC = "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599";
+
+      const result = await getTokenComponents(v2_yvWBTC, tokenEnsUri, protocolEnsUri, client);
+
+      expect(result.error).toBeFalsy();
+      expect(result.data).toBeTruthy();
+      expect(result.data).toMatchObject({
+        rate: "1",
+        unresolvedComponents: 0,
+        tokenAddress: v2_yvWBTC,
+        components: [
+          {
+            tokenAddress: WBTC,
+            components: [],
+            unresolvedComponents: 0,
+          },
+        ],
+      });
+      const tokenComponent = result.data as TokenComponent;
+      let sum = 0;
+      tokenComponent.components.forEach((x: TokenComponent) => {
+        sum += +x.rate;
+      });
+      expect(sum).toBeGreaterThan(0.95);
+      expect(sum).toBeLessThan(1.05);
```

Comment:
I don't think this will hold true for all the cases, yearn gives you yToken in exchange for the token (let's call it xToken) you submit to the vault, which will get used for liquidity mining and more xToken will be generated in vault and it will appreciate the price of yToken in terms of xToken. AFAIK, the rate here will most likely be greater than or equal to 1 otherwise you'd be losing the money by keeping your fund in the vault 😛. Although it would be nice to make sure this is the case.

Reply:
For sure, thanks. I adjusted the expected test case range to be between 0.95 and 1.15.

Reply:
At Roberto's request, I further increased the expected range to be between 1 and 100.